Add coordinator to emoncms

alexandrecuer opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Add coordinator to emoncms

Even if the emoncms API offers an endpoint per device, using the coordinator permits to update all the devices with a single request


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

alexandrecuer committed:
Add coordinator to emoncms

joostlek commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -84,19 +90,29 @@ async def async_setup_platform(
+            update_interval=scan_interval,
+        )
+    )
+    await coordinator.async_config_entry_first_refresh()
```
This is not using a config flow, so this function won't work as well. I know @zxdavb is also looking into this so maybe he has a pointer for this

alexandrecuer replied:
I've removed the `async_config_entry_first_refresh()` to use directly the client during initialization

Just tell me if acceptable

joostlek replied:
We should still use the coordinator to refresh, but I think you should `async_refresh` instead of the config entry one

zxdavb replied:
Using a **DataUpdateCoordinator** without Config Flow was quite straight-forward.

Assuming you have an awaitable `emoncms_client.async_list_feeds()`, your first step is as you have it:
```python
coordinator = DataUpdateCoordinator(
    hass,
    _LOGGER,
    name="emoncms",
    update_method=emoncms_client.async_list_feeds,
    update_interval=scan_interval,
)
```
... and to kick it off:
```python
await coordinator.async_config_entry_first_refresh() # get initial state
```

_HOWEVER, in between the two_, I had to add a fake listener, otherwise each refresh (including teh first, above) wouldn't schedule the next:
```python
# without a listener, DataUpdateCoordinator._schedule_refresh() won't be invoked by DataUpdateCoordinator._async_refresh()
coordinator.async_add_listener(lambda: None)
```

joostlek replied:
But my point being is that `await coordinator.async_config_entry_first_refresh()` isn't available for non-config entry integrations. I thought you had to deal with this with evohome

alexandrecuer replied:
uses now `async_refresh` and not `async_config_entry_first_refresh`

joostlek commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -84,19 +90,29 @@ async def async_setup_platform(
     emoncms_client = EmoncmsClient(url, apikey, session=async_get_clientsession(hass))
-    elems = await emoncms_client.async_list_feeds()
-
+    coordinator: DataUpdateCoordinator[list[dict[str, Any]] | None] = (
```
Let's create a new module `coordinator.py` where we create a full blown coordinator instead

alexandrecuer replied:
I thought it more concise this way
In found other integrations using a coordinator and without the coordinator.py 
cf the modbus one :
https://github.com/home-assistant/core/blob/1eb8b5a27c9f1c23c4ce8496c58c2b80394f2512/homeassistant/components/modbus/sensor.py#L91

joostlek replied:
We like to promote the consistency of placement of common patterns in their own module, and thus I think it would be nicer to have it in the `coordinator.py`

alexandrecuer replied:
Creation of a new module `coordinator.py` done

alexandrecuer committed:
Remove async_config_entry_first_refresh
As emoncms dont use config flow

alexandrecuer committed:
create a new module coordinator.py

joostlek commented on homeassistant/components/emoncms/coordinator.py:
```diff
@@ -0,0 +1,31 @@
+    """Emoncms Data Update Coordinator."""
+
+    def __init__(
+        self, hass: HomeAssistant, emoncms_client: EmoncmsClient, config: ConfigType
```
Let's only make the scan interval a parameter instead of the whole config

alexandrecuer replied:
Done

joostlek commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -115,47 +122,52 @@ async def async_setup_platform(
 
     def __init__(
         self,
         hass: HomeAssistant,
```
For a follow up, you don't need to set `self._hass` after the entity got added to home assistant, you can access the object in `self.hass`

alexandrecuer replied:
Thanks, so i remove as self._hass is not used actually in EmonCmsSensor...was there before, I forgot to remove

alexandrecuer committed:
dont pass to coordinator the whole config
But only the scan interval
plus some type hints corrected as there is a coordinator module now

frenck reviewed (approved):
Thanks, @alexandrecuer 👍 

../Frenck
